### PR TITLE
[CHANGED] MsgStore APIs now return errors

### PR DESCRIPTION
**Pull request by kozlovic**
This would allow the server to report possible failures.

### CODE REVIEW

File: server/monitor.go [hunk 1]
```diff
@@ -431,7 +431,7 @@ func (s *StanServer) handleOneChannel(w http.ResponseWriter, r *http.Request, na
 
 func updateChannelz(cz *Channelz, cs *stores.ChannelStore, subsOption int) {
 	msgs, bytes, _ := cs.Msgs.State()
-	fseq, lseq := cs.Msgs.FirstAndLastSequence()
+	fseq, lseq, _ := cs.Msgs.FirstAndLastSequence()
```
Comment (tylertreat) [inline server/monitor.go]: What is the reason for ignoring the error here?

File: server/server.go [hunk 2]
```diff
@@ -529,8 +531,10 @@ func (ss *subStore) Remove(cs *stores.ChannelStore, sub *subState, unsubscribe b
 			storageUpdate = sub.LastSent == qs.lastSent
 			sortedPendingMsgs := makeSortedPendingMsgs(sub.acksPending)
 			for _, pm := range sortedPendingMsgs {
-				m := cs.Msgs.Lookup(pm.seq)
-				if m == nil {
+				m, err := cs.Msgs.Lookup(pm.seq)
+				if m == nil || err != nil {
+					ss.stan.log.Errorf("Unable to update subscription for %s:%v (%v)",
+						sub.subject, pm.seq, err)
```
Comment (tylertreat) [inline server/server.go]: Why log here rather than bubbling the error up?

File: server/server.go [hunk 3]
```diff
@@ -529,8 +531,10 @@ func (ss *subStore) Remove(cs *stores.ChannelStore, sub *subState, unsubscribe b
 			storageUpdate = sub.LastSent == qs.lastSent
 			sortedPendingMsgs := makeSortedPendingMsgs(sub.acksPending)
 			for _, pm := range sortedPendingMsgs {
-				m := cs.Msgs.Lookup(pm.seq)
-				if m == nil {
+				m, err := cs.Msgs.Lookup(pm.seq)
+				if m == nil || err != nil {
```
Comment (tylertreat) [inline server/server.go]: In what situations is `m` nil and `err` not nil?

File: server/server.go [hunk 4]
```diff
@@ -3319,17 +3329,21 @@ func (s *StanServer) sendAvailableMessages(cs *stores.ChannelStore, sub *subStat
...
+		}
 		if nextMsg != nil {
 			return nextMsg
 		}
 		// Reason why we don't call FirstMsg here is that
 		// FirstMsg could be costly (read from disk, etc)
 		// to realize that the message is of lower sequence.
 		// So check with cheaper FirstSequence() first.
-		firstAvail := cs.Msgs.FirstSequence()
+		firstAvail, _ := cs.Msgs.FirstSequence()
```
Comment (tylertreat) [inline server/server.go]: Why ignore the error?

File: server/server.go [hunk 5]
```diff
@@ -3349,13 +3363,19 @@ func getNextMsg(cs *stores.ChannelStore, nextSeq, lastSent *uint64) *pb.MsgProto
 	}
 }
 
-func (s *StanServer) getSequenceFromStartTime(cs *stores.ChannelStore, startTime int64) uint64 {
-	return cs.Msgs.GetSequenceFromTimestamp(startTime)
+func (s *StanServer) getSequenceFromStartTime(cs *stores.ChannelStore, startTime int64) (uint64, error) {
+	tm, err := cs.Msgs.GetSequenceFromTimestamp(startTime)
+	if err != nil {
+		s.log.Errorf("Unable to get sequence from start time: %v", err)
+		return 0, err
```
Comment (tylertreat) [inline server/server.go]: Is this potential for double logging since we are logging here *and* propagating the error where it might be logged higher in the stack?

File: server/server.go (hunk 2 above)
Comment (kozlovic) [inline server/server.go]: We either need to panic/fatal, since we can't "move" this message to a different queue member, or log and continue. Same happens few lines below with the call to `AddSeqPending`.

File: server/server.go (hunk 4 above)
Comment (kozlovic) [inline server/server.go]: That will be in another PR where I need to handle gaps. For now simply ignore.

File: server/server.go (hunk 5 above)
Comment (kozlovic) [inline server/server.go]: I will, but the top layer may log something like "unable to add subscription: %v", and so this trace is adding more information: it happened when trying to get the sequence start. That message itself will contain the reason for the failure coming from the store. Unless we change from error to something that allows us to add causes (like a stack), we will always have this issue.

Comment (kozlovic) [inline server/server.go]: But ok, will remove. Actually, will remove that function. The caller of this one will call this directly and will bubble up the error to caller.

File: stores/common_test.go [hunk 6]
```diff
@@ -393,14 +452,14 @@ func testMaxMsgs(t *testing.T, s Store) {
 	}
 
 	// Check that older messages are no longer avail.
-	if cs.Msgs.Lookup(1) != nil || cs.Msgs.Lookup(uint64(firstSeqAfterLimitReached-1)) != nil {
+	if msgStoreLookup(t, cs.Msgs, 1) != nil || msgStoreLookup(t, cs.Msgs, uint64(firstSeqAfterLimitReached-1)) != nil {
```
Comment (petemiron) [inline stores/common_test.go]: suggest using a new line for each Or to assist in readability:
```
if msgStoreLookup(t, cs.Msgs, 1) != nil || 
     msgStoreLookup(t, cs.Msgs, uint64(firstSeqAfterLimitReached-1)) != nil {
```

File: stores/common_test.go [hunk 7]
```diff
@@ -758,10 +817,13 @@ func TestGSNoOps(t *testing.T) {
 	gms := &genericMsgStore{}
 	defer gms.Close()
 	gms.init("foo", testLogger, &limits.MsgStoreLimits)
-	if gms.Lookup(1) != nil || gms.FirstMsg() != nil || gms.LastMsg() != nil || gms.Flush() != nil ||
-		gms.GetSequenceFromTimestamp(0) != 0 || gms.Close() != nil {
+	if msgStoreLookup(t, gms, 1) != nil || msgStoreFirstMsg(t, gms) != nil || msgStoreLastMsg(t, gms) != nil ||
```
Comment (petemiron) [inline stores/common_test.go]: suggest multiline Or.

File: stores/common_test.go [hunk 8]
```diff
@@ -758,10 +817,13 @@ func TestGSNoOps(t *testing.T) {
...
+		gms.Flush() != nil || msgStoreGetSequenceFromTimestamp(t, gms, 0) != 0 || gms.Close() != nil {
```
Comment (petemiron) [inline stores/common_test.go]: suggest multiline Or.
